### PR TITLE
auto sync

### PR DESCRIPTION
**Pull request by murali-shris**
**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: at_client/lib/src/service/sync_service_impl.dart [hunk 1]
```diff
@@ -46,70 +48,106 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
-        // Setting isSyncInProgress to false, to allow next sync call.
-        _isSyncInProgress = false;
-        if (onError != null) {
-          onError(syncResult);
-        }
-        return syncResult;
+  void sync(
+      {Function? onDone,
+      Function? onError,
+      SyncRequestSource requestSource = SyncRequestSource.app}) {
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: We need not expose this. App by default. We will make it to system when it is system.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 2]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
-      return syncResult;
+  void sync({Function? onDone, Function? onError}) {
+    final syncRequest = SyncRequest();
+    syncRequest.onDone = onDone;
+    syncRequest.onError = onError;
+    syncRequest.requestedOn = DateTime.now().toUtc();
+    syncRequest.result = SyncResult();
+    _syncRequests.add(syncRequest);
+    //#TODO should we call _processQueue here or in cron or in _statsServiceListener ?
+    if (_syncRequests.length > _syncRequestThreshold ||
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: Not here. This should happen in a cron even when no one is calling sync.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 3]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
-      return await _sync(serverCommitId, onDone, onError);
-    } finally {
-      _isSyncInProgress = false;
+      return true;
+    } else if (syncRequest.result != null &&
+            syncRequest.result!.syncStatus == SyncStatus.failure ||
+        syncRequest.requestedOn.difference(DateTime.now().toUtc()).inMinutes >
+            _syncMaxRetryTime) {
+      if (syncRequest.onError != null) {
+        syncRequest.onError!(syncRequest.result);
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: I think onError need not be called. Will explain in a call.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 4]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
     }
-    try {
-      //Setting isSyncInProgress to true to prevent parallel sync calls.
-      _isSyncInProgress = true;
-      // If network is not available, return.
+    return;
+  }
+
+  /// Listens on stats notification sent by the cloud secondary server
+  Future<void> _statsServiceListener() async {
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:75]: Where is this method called?

File: at_client/lib/src/service/sync_service_impl.dart [hunk 5]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+        .listen((notification) async {
+      final serverCommitId = notification.value;
+      if (serverCommitId != null &&
+          int.parse(serverCommitId) > await _getLocalCommitId()) {
+        final syncRequest = SyncRequest();
+        syncRequest.onDone = _onDone;
+        syncRequest.onError = _onError;
+        syncRequest.requestSource = SyncRequestSource.system;
+        syncRequest.requestedOn = DateTime.now().toUtc();
+        syncRequest.result = SyncResult();
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:92]: Do we need result per request?

File: at_client/lib/src/service/sync_service_impl.dart [hunk 6]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+        _syncRequests.add(syncRequest);
+      }
+    });
+  }
+
+  void _processQueue() async {
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: This should be in a cron, and we don't need a for loop I think.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 7]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+
+  /// Listens on stats notification sent by the cloud secondary server
+  Future<void> _statsServiceListener() async {
+    _statsNotificationListener = await NotificationServiceImpl.create(_atClient)
+        as NotificationServiceImpl;
+    // Setting the regex to 'statsNotification' to receive only the notifications
+    // from stats notification service.
+    _statsNotificationListener
+        .subscribe(regex: 'statsNotification')
+        .listen((notification) async {
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:82]: Unrelated, but if any of the subscribers run to the exceptions NotificationService handing over Notifications should still live.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 8]
```diff
@@ -17,13 +17,17 @@ import 'package:at_lookup/at_lookup.dart';
...
 ///A [SyncService] object is used to ensure data in local secondary(e.g mobile device) and cloud secondary are in sync.
 class SyncServiceImpl implements SyncService, AtSignChangeListener {
-  bool _isSyncInProgress = false;
+  static const int _syncMaxRetryTime = 2,
+      _syncRequestThreshold = 3,
+      _syncRequestTriggerInSeconds = 30;
   late final AtClient _atClient;
   late final RemoteSecondary _remoteSecondary;
   late final NotificationServiceImpl _statsNotificationListener;
+  final _syncRequests = <SyncRequest>[];
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: Make it a fixed size queue. i.e 5 th one will be replaced.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 9]
```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+      final serverCommitId = notification.value;
+      if (serverCommitId != null &&
+          int.parse(serverCommitId) > await _getLocalCommitId()) {
+        final syncRequest = SyncRequest();
+        syncRequest.onDone = _onDone;
+        syncRequest.onError = _onError;
+        syncRequest.requestSource = SyncRequestSource.system;
+        syncRequest.requestedOn = DateTime.now().toUtc();
+        syncRequest.result = SyncResult();
+        _syncRequests.add(syncRequest);
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: Encapsulate in a method and these you decide whether to add or not and where to add

File: at_client/lib/src/service/sync_service_impl.dart (hunk 8 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

File: at_client/lib/src/service/sync_service_impl.dart (hunk 9 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

File: at_client/lib/src/service/sync_service_impl.dart (hunk 7 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart:82]: subscriber exception will not affect the notification service

File: at_client/lib/src/service/sync_service_impl.dart (hunk 6 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

File: at_client/lib/src/service/sync_service_impl.dart (hunk 4 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart:75]: inside create

File: at_client/lib/src/service/sync_service_impl.dart (hunk 3 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

File: at_client/lib/src/service/sync_service_impl.dart (hunk 2 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

File: at_client/lib/src/service/sync_service_impl.dart [hunk 10]
```diff
@@ -45,76 +53,117 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+    _logger.finer('system sync completed on ${syncResult.lastSyncedOn}');
+  }
+
+  void _onError(SyncResult syncResult) {
+    _logger.severe(
+        'system sync error ${syncResult.atClientException?.errorMessage}');
+  }
+
+  void _addSyncRequestToQueue(SyncRequest syncRequest) {
+    _syncRequests.removeLast();
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: This is a bug I think. This has be based on the size. I recon only one element will be present. Is it intentional?

File: at_client/lib/src/service/sync_service_impl.dart [hunk 11]
```diff
@@ -445,16 +474,19 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
     if (switchAtSignEvent.previousAtClient?.getCurrentAtSign() ==
         _atClient.getCurrentAtSign()) {
       // actions for previous atSign
+      _syncRequests.clear();
       _logger.finer(
           'stopping stats notificationlistener for ${_atClient.getCurrentAtSign()}');
       _statsNotificationListener.stopAllSubscriptions();
+      _cron.close();
+      _syncServiceMap.remove(_atClient.getCurrentAtSign());
     }
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:497]: You need an else here to start the cron. @X is switched to @Y and switched backed to @X is the scenario.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 12]
```diff
@@ -45,76 +53,117 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+        syncRequest.requestedOn = DateTime.now().toUtc();
+        _addSyncRequestToQueue(syncRequest);
       }
-      return syncResult;
+    });
+  }
+
+  void _processSyncRequests() async {
+    if (!await NetworkUtil.isNetworkAvailable()) {
+      _logger.finer('skipping sync due to network unavailability');
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:101]: You need a return here

File: at_client/lib/src/service/sync_service_impl.dart [hunk 13]
```diff
@@ -45,76 +53,117 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+    if (_syncRequests.length < _syncRequestThreshold &&
+        (_syncRequests.isNotEmpty &&
+            _syncRequests
+                    .elementAt(0)
+                    .requestedOn
+                    .difference(DateTime.now().toUtc())
+                    .inSeconds <
+                _syncRequestTriggerInSeconds)) {
+      _logger.finer(
+          'skipping sync - queue length ${_syncRequests.length} - first request time ${_syncRequests.elementAt(0).requestedOn}');
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: You need a return here

File: at_client/lib/src/service/sync_service_impl.dart [hunk 14]
```diff
@@ -45,76 +53,117 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
-        if (onError != null) {
-          onError(syncResult);
-        }
-        return syncResult;
+      if (await isInSync()) {
+        _logger.finer('server and local are in sync - ${syncRequest._id}');
+        syncRequest.result!
+          ..syncStatus = SyncStatus.success
+          ..lastSyncedOn = DateTime.now().toUtc();
+        _syncComplete(syncRequest);
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:128]: We have to call this on one of the app sync requests. The first one need not be an app sync request. If all of them are system sync requests then it is ok. We have to track the last app sync request separately or something.

File: at_client/lib/src/service/sync_service_impl.dart (hunk 11 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart:497]: i am clearing the sync service map ..when y is switched back to x..cron and stats listener will be started as a part of create

File: at_client/lib/src/service/sync_service_impl.dart (hunk 10 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: got it..will fix

File: at_client/lib/src/service/sync_service_impl.dart (hunk 12 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart:101]: done

File: at_client/lib/src/service/sync_service_impl.dart (hunk 13 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

File: at_client/lib/src/service/sync_service_impl.dart (hunk 14 above)
Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart:128]: will tackle this after initial round of testing

File: at_client/lib/src/service/sync_service_impl.dart (hunk 11 above)
Comment (VJag) [reply at_client/lib/src/service/sync_service_impl.dart:497]: Ok. Let's test the scenario once see if it is happening fine.

Comment (VJag) [reply at_client/lib/src/service/sync_service_impl.dart:497]: I still think you need an else to start the cron again

File: at_client/lib/src/service/sync_service_impl.dart [hunk 15]
```diff
@@ -45,76 +53,121 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+  }
+
+  void _onError(SyncResult syncResult) {
+    _logger.severe(
+        'system sync error ${syncResult.atClientException?.errorMessage}');
+  }
+
+  void _addSyncRequestToQueue(SyncRequest syncRequest) {
+    if(_syncRequests.length == _queueSize) {
+      _syncRequests.removeLast();
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart:169]: I hope there is some replaceLast. That would be more GC friendly.

File: at_client/lib/src/service/sync_service_impl.dart [hunk 16]
```diff
@@ -45,76 +53,121 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
     }
+    _syncRequests.addLast(syncRequest);
+    syncRequest.result = SyncResult();
```
Comment (VJag) [inline at_client/lib/src/service/sync_service_impl.dart]: ideally, this should happen elsewhere. This is a side effect. i.e doing beyond adding to queue.

Comment (murali-shris) [reply at_client/lib/src/service/sync_service_impl.dart]: done

**Review by sitaram-kalluri — APPROVED** [review]

**Review by naresh0689 — APPROVED** [review]

**Pull request merged**